rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,395 @@
+  - (-) It's hard to understand, type and implement, needs a special case for packages with few characters
+    - E.g. `x` could go in `pkgs/by-name/x-/--/x`
+  - (-) There's not enough packages even in Nixpkgs that a two-level 4-letter structure would make sense. Most of the structure would only be filled by a couple entries.
+  - (-) Even Git only uses 2-letter prefixes for its objects hex hashes
+- Use two-letter prefixes split into two directories, like `pkgs/by-name/h/e/hello`
+  - (+) Allows easy traversal by clicking on GitHub file listings, shard directories being limited to under 40 children
+  - (-) Requires special-casing single-letter attribute names
+    - (+) There's currently only 6 such cases, which could be handled on a one-off basis
+  - (-) Makes auto-completion worse, having to tab-complete once more
+  - (-) Makes it harder to create shards: if a shard doesn't exist yet, it has to be created with either one or two `mkdir`'s, or a `mkdir -p`
+- Use a dynamic structure where directories are rebalanced when they have too many entries.
+  E.g. `pkgs.foobar` could be in `pkgs/by-name/f/foobar` initially.
+  But when there's more than 1 000 packages starting with `f`, all packages starting with `f` are distributed under 2-letter prefixes, moving `foobar` to `pkgs/by-name/fo/foobar`.
+  - (-) The structure depends not only on the name of the package then, making it harder to find packages again and figure out where they should go
+  - (-) Complex to implement
+
+## Alternate `package.nix` filename
+
+Context: The only file that has to exist in package directories is `package.nix`, it must contain a function suitable for `callPackage`.
+
+Alternatives:
+- `default.nix`
+  - (+) `default.nix` is already a convention most people are used to.
+  - (-) We don't benefit from the usual `default.nix` benefits:
+    - Removing the need to specify the file name in expressions, but this does not apply because this file will be imported automatically by the code that replaces definitions from `all-packages.nix`.
+      - (+) But there's still some support for `all-packages.nix` for custom arguments, which requires people to type out the name
+        - (-) This is hopefully only temporary, in the future we should fully get rid of `all-packages.nix`
+    - Removing the need to specify the file name on the command line, but this does not apply because a package function must be imported into an expression before it can be used, making `nix build -f pkgs/by-name/hell/hello` equally broken regardless of file name.
+  - (-) Not using `default.nix` frees up `default.nix` for an expression that is actually buildable, e.g. `(import ../.. {}).hello`, although we don't yet have a use case for this that isn't covered by `nix-build ../.. -A <attrname>`.
+  - (-) Using `default.nix` would tempt users to invoke `nix-build .`, which wouldn't work and making package functions auto-callable is a known anti-pattern.
```

Comment:
I doubt people are actually doing that and this is something to worry about. Everyone I have observed and seen goes though all-packages.nix except when doing out of tree things which doesn't apply here anyway.

Reply:
> I doubt people are actually doing that and this is something to worry about. Everyone I have observed and seen goes though all-packages.nix except when doing out of tree things which doesn't apply here anyway.

The additional benefit of `package.nix` is that if one day nixos module files are put in the same dir as the package

Reply:
This RFC has been accepted & merged after extensive discussion several months ago, so I might be a bit late to add this input here. As I understand the project, the correct way would be to create a new RFC to suggest your desired change if you care strongly enough about it.

Reply:
Indeed, and you already brought this up before [here](https://github.com/NixOS/rfcs/pull/140#discussion_r1090915765)